### DISCUSSION TITLE
help with increasing convergence of a fiber reinforced soft actuator

### DISCUSSION THREAD

**Discussion by tolstoys19** [SofaDefrost/SoftRobots#212] [Help - Building / Configuring]
Dear SOFA Defrost community,

I am working on developing on a model of a small fiber reinforced soft actuator. For modeling the fiber reinforcement I am basing myself on this paper:

https://ieeexplore.ieee.org/abstract/document/9981637

I am using the stiffspringforcefield to model the fibers. The idea of the fibers is to limit radial expansion and also to assist with bending, all the loops are interconnected at the bottom section of the actuator(see BottomSection.jpeg).

I am currently having the problem that for high air pressures the actuator  looks like it explodes(see image ChamberProblem.jpg on the uploaded files) and stops converging. Do you have any ideas as to why this could be happening and any way to fix it? I have trying refining the mesh between coarser and finer sizes but it doesn't seem to help.

I have attached my files in here: The simulation is called measurements_working.py 

[QuestionConvergence.zip](https://github.com/SofaDefrost/SoftRobots/files/11261256/QuestionConvergence.zip)

PD: I would be grateful if you by any chance had any  feedback regarding the way I am modeling the fiber reinforcement.

Thank you for your time.

**Comment by EulalieCoevoet** [comment]
Hello, 

What I think happens is that the tetras of your mesh become flat (compressed between the cavity pressure and the fibers) and then start to flip. For instance it works better if you put the fibers on the cavity's wall.  

![image](https://user-images.githubusercontent.com/12150715/232792738-7d964daf-9f02-471a-9aef-62c468980dc5.png)

**Reply by tolstoys19** [reply]
Dear @EulalieCoevoet,

Thank you for your helpful I was able to increase the convergence doing what you said. I have some other  questions regarding this model. I would be very grateful if you could please take some time to answer them. 

 I measured the bending angle from the  model and compared it with experimental results(I attached a figure illustrating the angle I am measuring. Right now the results are quite different between the model and the experiment. In the model the actuator becomes increasingly easier to deform with higher air pressures(the tendency of the bending angle is always crescent) while in the experiment our actuator becomes harder to bend at high air pressures. Do you have any idea why this could be happening? I attached the bending angle vs input air pressure graph to this comment. I took some extra air pressures on the experiment just to show the tendency better,

 For now, my  main guess as to why this is happening is because we are using a linear material model and we might need a more complex one.  Even when using the hyperelastictetrahedronfemforcefield class all the models available are only order 1 models so there is not much of a change. I have seen in some papers(for example: https://ieeexplore.ieee.org/abstract/document/9981637) where do they a calibration of the young modulus. Would the solution be to do a similar thing for my model or do you have any other recommendation?

My actuator is made of dragonskin10 medium.

Thank you for your time
![bendingangle](https://user-images.githubusercontent.com/84059736/233077457-5e23a700-070c-4973-893e-dc629ddeb51a.jpg)

![image](https://user-images.githubusercontent.com/84059736/233077793-11351be3-d264-41a1-baf8-c6e56de04a68.png)

**Reply by EulalieCoevoet** [reply]
I think you need more dofs, you don't have enough points to allow your model to deform between each fiber. I know you said you already tried, but maybe this could solve the original problem.
Also I think that you should set a higher Poisson's ratio, I don't think 0.3 is enough. Silicone is not compressible, maybe that's why it is easier to bend in simulation. Also, is it possible that at some point the bending creates auto collisions on your prototype ?

**Reply by tolstoys19** [reply]
@EulalieCoevoet Thank you for your quick response. I did some of the changes you suggested:
I refined the meshes for both the chamber and the actuators body. The maximum size I allowed per element is of 1mm(previously set to 2 mm) but the problem is that the simulation becomes very heavy. I am running this on an Intel(R) Core(TM) i7-10750H CPU with 16GB of RAM and the simulation is running at around 1fps.  Do you know of any way I could optimize this?

I uploaded the files with the new settings. In the data/mesh folder you can find the step files for the actuator(actbodywithconnectoredit.stp) and the chamber(CHAMBER.stp) in case you know any better way to mesh them.  Right now I am using GMSH I set the maximum element size to 1mm. For the actuators body I am using the fragment boolean to connect the separate volumes.
 I also set a Poisson Ratio of 0.49 but I don't see much changes in the tendency. The young modulus I am using right now is a sample value but the main problem is the change in tendency. The experiment is much easier to bend  below around 80kPa than the model  and then gradually becomes harder to bend. The model doesn't match this tendency(the model gradually becomes easier and easier to deform). Regarding the self collisions, the picture I uploaded in the previous comment is the prototype at maximum air pressure. As you can see there is  inflation between the fiber loops and there is a possibility of small self collisions happening in between this sections  but I don't know if it's really something that significant. 


Thank you for helping so much.
[Question2.zip](https://github.com/SofaDefrost/SoftRobots/files/11280112/Question2.zip)
